From be680de5b8e0fe681b80f8b408e0715ad15817be Mon Sep 17 00:00:00 2001
From: Yuvarani V <quic_yuvarani@quicinc.com>
Date: Tue, 12 Sep 2023 14:53:32 +0530
Subject: [PATCH] hostapd: Add support for MLD ID Probing

Add support for Probing MLD ID corresponding to NonTxBSSID
with A1 set to TxBSSID of the MLD.

Signed-off-by: Yuvarani V <quic_yuvarani@quicinc.com>
---
 src/ap/beacon.c                | 13 +++++++++++++
 src/ap/ieee802_11.c            | 10 +---------
 src/common/ieee802_11_common.h |  1 +
 3 files changed, 15 insertions(+), 9 deletions(-)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index 7062ef7..082ac6c 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -1708,6 +1708,19 @@ void handle_probe_req(struct hostapd_data *hapd,
 		hostapd_parse_mle(hapd, WLAN_FC_STYPE_PROBE_REQ,
 				  &elems, &ml_data);
 
+	/* MLD ID Probing */
+	if ((hapd == hostapd_mbssid_get_tx_bss(hapd)) &&
+	    (ml_data.u.preq.mld_id != 0)) {
+		if (ml_data.u.preq.mld_id < hapd->iface->num_bss)
+			hapd = hapd->iface->bss[ml_data.u.preq.mld_id];
+		else {
+			wpa_printf(MSG_MSGDUMP, "Ignore Probe Request from " MACSTR
+				   " since No Matched Non-tx vap found for BSSID Index %d",
+				   MAC2STR(mgmt->sa), ml_data.u.preq.mld_id);
+			return;
+		}
+	}
+
 	wpa_msg_ctrl(hapd->msg_ctx, MSG_INFO, RX_PROBE_REQUEST "sa=" MACSTR
 		     " signal=%d", MAC2STR(mgmt->sa), ssi_signal);
 
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index eb967f6..973ae9e 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -9671,15 +9671,7 @@ static void hostapd_parse_probe_req_mle(struct hostapd_data *hapd,
 			return;
 		}
 		mld_id = *(mle_pos + 1);
-
-		/* Handle ml probe requests to self AP MLD
-		 * TODO MBSS case
-		 */
-		if (mld_id != 0) {
-			wpa_printf(MSG_ERROR, "MLD ID mismatch in ML Probe request mld id %d", mld_id);
-			ml_data->present = false;
-			return;
-		}
+		ml_data->u.preq.mld_id = mld_id;
 	}
 
 	mle_pos += cmn_info_len;
diff --git a/src/common/ieee802_11_common.h b/src/common/ieee802_11_common.h
index eb1d16e..d252656 100644
--- a/src/common/ieee802_11_common.h
+++ b/src/common/ieee802_11_common.h
@@ -99,6 +99,7 @@ struct multi_link_preq_info {
 	 * link_info struct is not valid
 	 */
 	u16 link_bmap;
+	u8 mld_id;
 	struct {
 		bool complete;
 	} link_info[MAX_SUPPORTED_LINKS];
-- 
2.17.1

